File: runtimes/swiftui/Sources/StyleConverterRuntime/Renderer/ComponentHost.swift
//
//  ComponentHost.swift
//  Renderer — IR v2 Slot & Placement contract (design §3.2).
//
//  The runtime shim between composition and the engine: EVERY component
//  renders through ComponentHost, which unconditionally attaches the
//  component's self-extracted ItemPlacement as a SwiftUI layout value.
//  Parent-data is only read by whoever MEASURES the node, so the
//  placement is inert under a plain VStack/ZStack and active under
//  CSSGridLayout/CSSFlexLayout — the child never knows its destination,
//  the container never knows its arrivals until measure time.
//
//  Note on the design's `.zIndex(…)` mention: paint order is already
//  applied by the component's own PositionApplier inside
//  ComponentRenderer.body (a ZIndex declaration acts among siblings
//  wherever the view lands), so the host does NOT apply it a second
//  time — doubling the modifier would change the campaign-pinned paint
//  order. The claim still rides ItemPlacement.paint for introspection.
//

import SwiftUI

/// Composition-side wrapper: ComponentRenderer + placement parent-data.
/// Use this (not ComponentRenderer directly) wherever a component enters
/// a view tree — the renderer's child loop, the harness gallery and the
/// capture canvas all route through it.
// public: the harness (CaptureCanvas, ComponentGallery) hosts root
// components through this shim, same as the engine hosts children.
public struct ComponentHost: View {
    /// The component to render — placement is derived from ITS OWN
    /// properties only (no parent participation, design §2.2).
    let component: IRComponent

    // public: explicit memberwise init — the synthesized one is
    // internal, so cross-module callers (the harness) need it spelled out.
    public init(component: IRComponent) {
        self.component = component
    }

    // public: View protocol witness on a public type must be public.
    public var body: some View {
        ComponentRenderer(component: component)
            // The v2 parent-data channel: ITEM-scoped declarations
            // (align-self, flex-*, grid placement, z-index — the frozen
            // PropertyScope.item list) packaged for whatever container
            // this component lands in.
            .stylePlacement(ItemPlacementExtractor.extract(from: component.properties))
    }
}

File: runtimes/compose/src/main/java/com/styleconverter/runtime/core/renderer/ComponentHost.kt
package com.styleconverter.runtime.core.renderer

// ComponentHost — the runtime shim of the v2 contract (design §3.1): the
// single entry point a composer (harness screen or SDUI shell) uses to
// render ONE component. For EVERY component it derives the ITEM-scoped
// placement union once and attaches it as Compose parent-data
// (Modifier.itemPlacement) before delegating to the style engine's
// ComponentRenderer. Parent-data is only read by whoever measures the
// node, so the claim is inert under a plain host Box and active under a
// placement-aware container — the child never knows its destination.
//
// The host rides RenderComponent's existing itemModifier channel (the
// outermost slot of the component's own modifier chain), adding ZERO new
// layout nodes — a parent-data modifier has no measure/draw effect of its
// own, which is what keeps this wiring pixel-identical to the pre-v2
// render path (the campaign baseline freeze).

import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import com.styleconverter.runtime.core.ir.IRComponent
import com.styleconverter.runtime.core.placement.ItemPlacementExtractor
import com.styleconverter.runtime.core.placement.itemPlacement

object ComponentHost {

    /**
     * Render one component with its placement claims published as
     * parent-data. Children arrive already composed on the component
     * (SlotComposer output) — the renderer walks them as slot content.
     */
    @Composable
    fun Render(component: IRComponent) {
        // css-cascade-5 §7.3.2 ("Explicit Inheritance: the inherit
        // keyword") makes `background-color: inherit` compute to the
        // INHERITED VALUE, which §7.2 defines as the computed value of the
        // property on the parent element — something no extractor can see
        // (it is handed one component's declarations). Resolve it here, once, over the whole
        // composed subtree — this shim is the single root entry point, and
        // the renderer walks the children off the component it is given, so
        // one pass covers the tree. Identity pass-through for any tree that
        // does not use the keyword. See BackgroundColorInheritance.
        val prepared = remember(component) {
            com.styleconverter.runtime.color.BackgroundColorInheritance.resolve(component)
        }
        // Extract once per component identity — the claims are pure
        // functions of the IR properties.
        val placement = remember(prepared) {
            ItemPlacementExtractor.extract(prepared.properties)
        }
        ComponentRenderer.RenderComponent(
            prepared,
            // itemModifier is prepended OUTERMOST in RenderComponent's
            // style chain, so the parent-data lands on the component's
            // root layout node — the measurable its container sees.
            itemModifier = Modifier.itemPlacement(placement)
        )
    }
}
